tools/lib/ensure-coreutils.sh

```diff
@@ -4,17 +4,34 @@
 # Ensures GNU coreutils are available in PATH.
 # May add to PATH, or exit with a message to stderr.
 #
-# We don't yet have good docs for handling failure here if on
-# Windows or macOS, so until then this isn't suitable for scripts
-# used in normal development -- only for maintainer-facing scripts.
+# The GNU coreutils are a basic piece of infrastructure for having
+# a reasonable 21st-century shell scripting environment, so we
+# freely invoke this in any of our scripts that need `readlink -f`
+# or other features not always found without them.
+#
+# On any GNU/Linux system this is of course a non-issue, and nor
+# is it on Windows: we inevitably require Git, and Git for Windows
+# comes with a GNU environment called "Git BASH", based on MSYS2.
+#
+# So this is really all about macOS.  Fortunately it's easy to get
+# coreutils installed there too... plus, many people already have
+# it installed but just not in their PATH.  We write our scripts
+# for a GNU environment, so we bring it into the PATH.
 
 check_coreutils() {
-    fmt --help >/dev/null 2>&1
+    # Check a couple of commands for GNU-style --help and --version,
+    # which macOS's default BSD-based implementations don't understand.
+    command fmt --help >/dev/null 2>&1
+    command readlink --version 2>&1 | grep -q GNU
 }
 
 ensure_coreutils() {
+    # If we already have it, then great.
     check_coreutils && return
 
+    # Homebrew provides names like `greadlink` on the PATH,
+    # but also provides the standard names in this directory.
+    # Try that.
     homebrew_gnubin=/usr/local/opt/coreutils/libexec/gnubin
     if [ -d "$homebrew_gnubin" ]; then
         export PATH="$homebrew_gnubin":"$PATH"
```

Comment:
This is technically incorrect: `$PATH` may be empty. :upside_down_face:

Reply:
Hmmm -- I guess that can be possible :stuck_out_tongue: but it seems like a fairly pathological situation to be in.

Do you have an idiom for smoothly handling that situation? I am not sure I've ever seen a script that took account of that possibility. Could handle it if it doesn't add much complication... but I'm inclined to have a low threshold on that level of complexity. We also assume things like: you don't have a `readlink` that is actually some version from the '80s and lacks a feature like `-m`.

Reply:
I believe replacing the hard colon with `${PATH:+:}` would do it.

(Not worth the trouble here, mind, but it might be useful in other contexts.)

Reply:
> (Not worth the trouble here, mind, but it might be useful in other contexts.)

:+1: